lec07-pop-models.Rmd

```diff
@@ -1,4 +1,617 @@
+  r <- parameters[1] # the first element of the parameters vector is r
+  K <- parameters[2] # the second element of the parameters vector is K
+
+  #rate of change
+  dN = r*N*(1-N/K)
+    
+  #return rate of change
+  list(c(dN))
+}
+
+parameters <- c(r=0.5,K=50)
+state <- c(N=10)
+times <- seq(0, 50, by = 0.01)
+
+help(ode) # look at the documentation to learn about the parameters
+result <- ode(y = state, times = times, func = logistic_fn, parms = parameters)
+```
+
+The output of `ode` is a vector that contains the times we requested and its calculated values of N. 
+
+Notice that by using `ode` we didn't have to write a loop explicitly ourselves, and we didn't have to
+choose a timestep. `ode` does under the hood a version of what we did explicitly before. 
+
+```{r}
+print(head(result,10)) # result has a list of times and values of N at those times
+print(typeof(result)) # result is a 'double': a vector of numbers.
+```
+
+```{r}
+
+result <- data.frame(result) # convert it to a dataframe so we can use ggplot
+
+ggplot(result) +
+  geom_point(aes(x = time, y = N))
+```
+
+Now let's make a nice plot of numerical solutions from a few different starting points.
+
+```{r}
+library(dplyr)
+
+parameters <- c(r=0.5,K=50)
+initial_conditions <- seq(10, 80, by = 10)
+times <- seq(0, 50, by = 0.01)
+
+results <- list() # make an empty list
+count <- 1
+
+for (N0 in initial_conditions) {
+  
+  result <- ode(y = N0, times = times, func = logistic_fn, parms = parameters)
+  result <- data.frame(result)
+  
+  results[[count]] <- result
+  
+  count <- count + 1
+}
+
+all_results <- bind_cols(results)
+```
+
+```{r}
+View(all_results) # look at what we got
+
+print(colnames(all_results)) # column names
+```
+
+
+
+```{r}
+ggplot(all_results) +
+  geom_line(aes(x = time, y = X1)) +
+  geom_line(aes(x = time1, y = X11)) +
+  geom_line(aes(x = time7, y = X17)) # etc, we can keep adding more here
+
+```
+
+
+
+## Qualitative analysis of 2D models
+ - fixed points
+ - stability
+ - phase portraits
+ - introduce null clines - *cline* comes from a Greek word meaning "slope"
+ - introduce isoclines? null cline is a special case of isocline
+
+
+
+```{r}
+library(deSolve)
+lotka_volterra <- function(t,y,parameters) { 
+  # the arguments need to be named "t", "y", and "parameters", otherwise this won't work with phaseR
+    
+    X <- y[1]; Y <- y[2]
+    
+    alpha <- parameters[1]; beta <- parameters[2]; gamma <- parameters[3]; delta <- parameters[4]
+    
+    print(X)
+    print(Y)
+    
+    #rate of change
+    dx <- alpha*X - beta*X*Y
+    dy <- delta*X*Y - gamma*Y
+    
+    #return rate of change
+    list(c(dx, dy))
+}
+```
+
+```{r}
+lotka_volterra(2,y = c(0.6,0.5), parameters = c(0.5,0.6,0.7,0.8))
+```
+
+
+```{r}
+# deSolve with lotka volterra
+
+#parameters = c(alpha = 0.5, beta = 0.6, gamma = 0.5, delta = 0.5)
+
+parameters = c(0.5, 0.6, 0.9, 0.5)
+
+state <- c(X = 0.5,
+ Y = 0.5) # these variables need to match the function arguments in label
```

Comment:
Why new line here?

Reply:
I'm stopping at this point since I am guessing you are still working on this bit.

Reply:
Yep this part isn't done yet.